listenbrainz/mbid_mapping/reports/top_discoveries.py

```diff
@@ -0,0 +1,135 @@
+        raise
+
+
+def create_indexes(mb_conn):
+    """ Create the user_name index on the top discoveries table. """
+
+    try:
+        with mb_conn.cursor() as curs:
+            curs.execute("""CREATE INDEX top_discoveries_ndx_user_name
+                                      ON mapping.top_discoveries (user_name)""")
+            mb_conn.commit()
+    except (psycopg2.errors.OperationalError, psycopg2.errors.UndefinedTable) as err:
+        log("mbid mapping: failed to create top discoveries indexes", err)
+        mb_conn.rollback()
+        raise
+
+
+def fetch_user_list(lb_conn, year):
+    """ Fetch the active users in the given year """
+
+    with lb_conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as lb_curs:
+        query = """SELECT DISTINCT user_name
+                     FROM listen
+                    WHERE listened_at >= %s
+                      AND listened_at < %s"""
+
+        start_ts = int(datetime(year, 1, 1, 0, 0, tzinfo=timezone.utc).timestamp())
+        end_ts = int(datetime(year + 1, 1, 1, 0, 0, tzinfo=timezone.utc).timestamp())
+        lb_curs.execute(query, (start_ts, end_ts))
+        rows = lb_curs.fetchall()
+
+    return [r[0] for r in rows]
+
+
+def chunks(lst, n):
+    """ Helper function to break a list into chunks """
+    for i in range(0, len(lst), n):
+        yield lst[i:i + n]
+
+
+def fetch_top_discoveries_for_users(lb_conn, mb_conn, year):
+    """ Actually fetch the top discoveries for the given year"""
+
+    with lb_conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as lb_curs:
+        with mb_conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as mb_curs:
+            log("crate top_listens table")
+            create_table(mb_conn)
+
+            log("fetch active users")
+            user_list = fetch_user_list(lb_conn, year)
+            log("Process %d users." % len(user_list))
+
+            query = """SELECT user_name
+                            , track_name
+                            , data->'track_metadata'->>'artist_name' AS artist_name
+                            , array_agg(extract(year from to_timestamp(listened_at))::INT ORDER BY
+                                        extract(year from to_timestamp(listened_at))::INT) AS years
+                            , m.recording_mbid
+                            , mm.artist_mbids
+                         FROM listen
+              FULL OUTER JOIN mbid_mapping m
+                           ON (data->'track_metadata'->'additional_info'->>'recording_msid')::uuid = m.recording_msid
+              FULL OUTER JOIN mbid_mapping_metadata mm
+                              ON mm.recording_mbid = m.recording_mbid
+                        WHERE user_name in %s
+                          AND mm.recording_mbid IS NOT NULL
+                     GROUP BY user_name, artist_name, mm.artist_mbids, track_name, m.recording_mbid
+                       HAVING (array_agg(extract(year from to_timestamp(listened_at))::INT ORDER BY
```

Comment:
I think this could become `min(extract(year from to_timestamp(listened_at))::INT) = %s` and then the array_agg in select cause could become just a `count(*)`.

Reply:
I find that PG is not entirely predictable for when this works, but it doesn't always work. I'm going to call this good enough and leave it be.